src/bootstrap/builder.rs

```diff
@@ -210,14 +229,14 @@ impl StepDescription {
                 Err(_) => path,
             };
 
-            let mut attempted_run = false;
-            for (desc, should_run) in v.iter().zip(&should_runs) {
-                if let Some(suite) = should_run.is_suite_path(path) {
-                    attempted_run = true;
-                    desc.maybe_run(builder, suite);
-                } else if let Some(pathset) = should_run.pathset_for_path(path) {
-                    attempted_run = true;
-                    desc.maybe_run(builder, pathset);
+            let mut attempted_run = match_run(v, builder, &should_runs, path);
+            if !attempted_run {
+                // If the path is like this: compiler/rustc_data_structures/src/...
+                // Then get rid of the src/... part and try again
```

Comment:
Instead of this splitting (which I don't think works ideally on e.g. src/foo paths, where we'll try to look for an empty path... I think it may be enough to add a ShouldRun::krate variant that adds each crate as a suite path instead of a set -- that would give us prefix detection "for free".

Reply:
So should I change `krate` and `all_krates` to do this or should I add separate functions that do this? I am not sure, for the latter, what crates should use the `krate_suite` variant.